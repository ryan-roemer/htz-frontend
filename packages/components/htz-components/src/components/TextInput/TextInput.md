### TextInput

`<TextInput>` is a styleable, easy to use and a11y compliant text input component.  
It can be used as a single/multi line simple text input or as a rich text editor that has Bold and Italic format buttons.

**React Controllable Input**

**From the react docs:**  
In most cases, we recommend using controlled components to implement forms. In a controlled component, form data is handled by a React component. The alternative is uncontrolled components, where form data is handled by the DOM itself.

### Features

* Controllable, can be used both as controlled and uncontrolled input
* single and multiline support
* simple and rich text editing
* a11y compliant
* easly styled with variants

### Uncontrolled Examples

**Minimum required props**

At minimum can be used with just a label prop which is required for a11y

```jsx
<div dir="rtl">
  <TextInput label="label" />
</div>
```

**Hidden label**

`label` is a required prop for a11y compliance, if its not visually needed a `labelHidden` prop can be passed, this will add an `aria-label` attr to the relavent element with the label text, and hide the visual label.

```jsx
<div dir="rtl">
  <TextInput label="label" labelHidden />
</div>
```

**Required Input**

Used to Indicate the input is required.

When passed as a prop, the following component is rendered after the label:

See documantation of `requiredText` prop above to see how to manipulate the styling and presentation of the `<abbr/>`

```jsx static
<abbr className={className} title={requiredText.long}>{' '}
 {isSup ? <sup>{requiredText.short}</sup> : requiredText.short}
</abbr>
```

* Default color is dependant on the variant
* Custom styling can be passed with miscStyles

```jsx
<div dir="rtl">
  <TextInput
    label="label"
    requiredText={{
      long: "this field is required",
      short: "*"
    }}
  />
</div>
```

**Color Variants**

An application theme may define stylistic variants for inputs to be used in different UI situations.
Picking a variant changes multiple aspects of the `<TextInput />`'s styling at once.

The `primaryInverse` variant is intended for use when placed on dark backgrounds

```jsx static
  <TextInput variant="primaryInverse"... />
```

```jsx
<div dir="rtl" style={{ backgroundColor: "black", padding: "5rem" }}>
  <TextInput variant="primaryInverse" label="label" noteText="inverse note" />
</div>
```

**Dom Element Attributes**

The `attrs` prop take an object which is spread on the DOM element, and can be used to add additional attributes to the DOM element, e.g.:

```jsx static
  <TextInput
    attrs={{ "aria-label": "A input with custom attributes" }}...
  />
```

```jsx
<div dir="rtl">
  <TextInput
    label="label"
    attrs={{ "aria-label": "A input with custom attributes" }}
  />
</div>
```

**Default Value**

Used to specify the initial value only in uncontrolled `<TextInput />`'s

```jsx
<div dir="rtl">
  <TextInput label="label" defaultValue="Im the default value" />
</div>
```

**Error state**

When the `isError` prop is provided, the styling will be changed accordingly to affect settings in the theme.
Additionally, when `errorText` is provided it will replace the `noteText` prop

```jsx
<div dir="rtl">
  <TextInput isError label="דואל" errorText="אנא הזינו כתובת תקינה" />
</div>
```

**Example of Changing to Error state**

Notice that when adding isError prop the errorText will replace the noteText

The Input type can be controlled with the type prop

```jsx
<div dir="rtl">
  <TextInput
    label="דואל"
    noteText="נא להכניס דואל"
    errorText="נא להכניס כתובת דואל תקינה"
    placeHolder="example@you.com"
    type="email"
  />
  <TextInput
    label="דואל"
    noteText="נא להכניס דואל"
    isError
    errorText="נא להכניס כתובת דואל תקינה"
    placeHolder="example@you.com"
    type="email"
  />
</div>
```

**Disabled Input**

```jsx
<div dir="rtl">
  <TextInput label="label" isDisabled />
</div>
```

**Max and Min length**

Passing `maxLength` and `minLength` props will add native html max and min length attributes, limiting the number of characters.

Notice these wont work when isContentEditable is passed (rich text editing)

```jsx
<div dir="rtl">
  <TextInput label="דואל" maxLength={6} minLength={2} />
</div>
```

**Focus and Blur callbacks**

Triggered when the `<InputElement />` gets focused/blured

(check the console to see logs)

```jsx static
  <TextInput
    onFocus={() => console.log("input focused")}
    onBlur={() => console.log("input blured")}
    ...
  />
```

```jsx
<div dir="rtl">
  <TextInput
    label="דואל"
    noteText="נא להכניס דואל"
    errorText="נא להכניס כתובת דואל תקינה"
    placeHolder="example@you.com"
    type="email"
    onFocus={() => console.log("input focused")}
    onBlur={() => console.log("input blured")}
  />
</div>
```

**Multi-line simple text editing**

Pass the `isTextArea` prop

```jsx
<div dir="rtl">
  <TextInput
    noteText="אנא הזינו כתובת "
    errorText="אנא הזינו כתובת תקינה"
    label="דואל"
    placeholder="כתובת מייל"
    isTextArea
  />
</div>
```

**Multi-line custom height**

When using multiline, (textarea or contenteditable), a `height` prop can be passed.
`height` can be a simple number (e.g `height={5}`), or an array of responsive options objects.

The number value should be the number of rems the element height should equal.

```jsx
<div dir="rtl">
  <TextInput
    noteText="אנא הזינו כתובת "
    errorText="אנא הזינו כתובת תקינה"
    label="דואל"
    placeholder="כתובת מייל"
    isTextArea
    height={[{ until: "s", value: 10 }, { from: "l", value: 23 }]}
  />
</div>
```

**Rich Text Editing with ContentEditable**

Passing `isContentEditable` will enable rich text editing, with bold and Italic formating buttons.

```jsx
<div dir="rtl">
  <TextInput
    noteText="אנא הזינו כתובת "
    errorText="אנא הזינו כתובת תקינה"
    label="דואל"
    placeholder="כתובת מייל"
    isContentEditable
  />
</div>
```

Disabling a contenteditable requires custom logic, `<TextInput />` takes care of that logic and passing `isDisabled` prop works with a contenteditable just like input and textarea.

```jsx
<div dir="rtl">
  <TextInput
    noteText="אנא הזינו כתובת "
    errorText="אנא הזינו כתובת תקינה"
    label="דואל"
    placeholder="כתובת מייל"
    isContentEditable
    isDisabled
  />
</div>
```

### Conrolled Examples

**Basic usage with minimum required props**

Simple example of changing state `onChange` of `<TextInput />`, and sending the value down creates a React controlled Input.

```jsx static
class Example extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: ""
    };
  }

  render() {
    return (
        <div>the value from the state is: {this.state.value}</div>
        <TextInput
          label="label"
          onChange={evt => this.setState({ value: evt.target.value })}
          value={this.state.value}
        />;
    );
  }
}
  <Example />
```

```jsx
class Example extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: ""
    };
  }

  render() {
    return (
      <div dir="rtl">
        <div>the value from the state is: {this.state.value}</div>
        <TextInput
          label="label"
          onChange={evt => this.setState({ value: evt.target.value })}
          value={this.state.value}
        />;
      </div>
    );
  }
}
<div dir="rtl">
  <Example />
</div>;
```

**onInput vs onChange**

When not using with `isContentEditable` prop, either `onChange` or `onInput` callbacks can be used according to the use case.

```jsx static
        <div>the value from the state is: {this.state.value}</div>
        <TextInput
          label="label"
          isTextArea
          height={15}
          onInput={evt => this.setState({ value: evt.target.value })}
          value={this.state.value}
        />;
```

```jsx
class Example extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: ""
    };
  }

  render() {
    return (
      <div dir="rtl">
        <div>the value from the state is: {this.state.value}</div>
        <TextInput
          label="label"
          isTextArea
          height={15}
          onInput={evt => this.setState({ value: evt.target.value })}
          value={this.state.value}
        />;
      </div>
    );
  }
}
<div dir="rtl">
  <Example />
</div>;
```

**Rich Text controlled usage**

When using as a RichText editor `onContentEditableChange` callback prop should be used, this callback gets the value as the second parameter.  
The value is the innerHtml of the contenteditable div

```jsx static
        <div>the value from the state is: {this.state.value}</div>
        <TextInput
          label="label"
          isContentEditable
          height={10}
          onContentEditableChange={(evt, value) => this.setState({ value })}
          value={this.state.value}
        />;
      </div>
```

```jsx
class Example extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: ""
    };
  }

  render() {
    return (
      <div dir="rtl">
        <div>the value from the state is: {this.state.value}</div>
        <TextInput
          label="label"
          isContentEditable
          height={10}
          onContentEditableChange={(evt, value) => this.setState({ value })}
          value={this.state.value}
        />;
      </div>
    );
  }
}
<div dir="rtl">
  <Example />
</div>;
```

**Getting a ref on the Input**

A `refFunc` callback prop can be passed in order to get a ref of the input/textarea/contenteditable.
An example use case is needing to call .focus() on the element  
Refs should be used carefully, as explained in the react docs: http://bit.ly/2BJh1KA

```jsx static
        <div>the value from the state is: {this.state.value}</div>
        <Button
          onClick={() => this.elem.focus()}
          miscStyles={{ marginTop: "3rem" }}
        >
          focus on element
        </Button>
        <TextInput
          label="label"
          refFunc={el => (this.elem = el)}
          onChange={evt => this.setState({ value: evt.target.value })}
          value={this.state.value}
          miscStyles={{ marginTop: "3rem" }}
        />;
```

```jsx
class Example extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: ""
    };
  }

  render() {
    return (
      <div dir="rtl">
        <div>the value from the state is: {this.state.value}</div>
        <Button
          onClick={() => this.elem.focus()}
          miscStyles={{ marginTop: "3rem" }}
        >
          focus on element
        </Button>
        <TextInput
          label="label"
          refFunc={el => (this.elem = el)}
          onChange={evt => this.setState({ value: evt.target.value })}
          value={this.state.value}
          miscStyles={{ marginTop: "3rem" }}
        />;
      </div>
    );
  }
}
<div dir="rtl">
  <Example />
</div>;
```
