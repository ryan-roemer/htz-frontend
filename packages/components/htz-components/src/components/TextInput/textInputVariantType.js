import { oneOf, } from 'prop-types';

export const textInputVariantType = oneOf([
  'primary',
  'primaryInverse',
  'search',
]);
