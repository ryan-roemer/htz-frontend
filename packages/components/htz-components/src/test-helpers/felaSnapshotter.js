import React from 'react';
import createRenderer from '../components/StyleProvider/createRenderer';
import StyleProvider from '../components/StyleProvider/StyleProvider';
import createFelaSnapshotter from './createFelaSnapshotter';

const testRenderer = createRenderer({ isRtl: true, });
const felaSnapshotter = createFelaSnapshotter(StyleProvider, testRenderer);

export default felaSnapshotter;
