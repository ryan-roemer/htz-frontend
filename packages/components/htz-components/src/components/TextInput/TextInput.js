/* global document */

import React, { Component, } from 'react';
import PropTypes from 'prop-types';
import { createComponent, } from 'react-fela';
import {
  border,
  borderBottom,
  parseStyleProps,
  parseComponentProp,
} from '@haaretz/htz-css-tools';
import { attrsPropType, } from '../../propTypes/attrsPropType';
import { stylesPropType, } from '../../propTypes/stylesPropType';
import Abbr from './elements/Abbr';
import Button from '../Button/Button'; // eslint-disable-line import/no-named-as-default
import IconBold from '../Icon/icons/IconBold';
import IconItalic from '../Icon/icons/IconItalic';
import InputElement from './elements/InputElement';
import Note from '../Note/Note';
import { responsivePropBaseType, } from '../../propTypes/responsivePropBaseType';
import { textInputVariantType, } from './textInputVariantType';

const labelStyle = ({
  boxModel,
  isContentEditable,
  isDisabled,
  isError,
  isFocused,
  isTextArea,
  isLabelUp,
  miscStyles,
  theme,
  variant,
}) => ({
  cursor: 'pointer',
  display: 'flex',
  alignItems: 'center',
  ...(isTextArea || isContentEditable
    ? { flexDirection: 'column', alignItems: 'flex-start', }
    : {}),
  paddingInlineStart: `${boxModel && boxModel.hp ? boxModel.hp : 1}rem`,
  paddingInlineEnd: `${boxModel && boxModel.hp ? boxModel.hp : 1}rem`,
  width: '100%',
  ...(!isTextArea && !isContentEditable
    ? { position: 'relative', marginTop: '1rem', }
    : {}),
  extend: [
    theme.type(theme.inputStyle.typeScale),
    parseComponentProp(
      undefined,
      variant,
      theme.mq,
      setVariant,
      theme.color,
      boxModel,
      isError,
      isFocused,
      theme,
      isDisabled,
      isLabelUp
    ),
    // Trump all other styles with those defined in `miscStyles`
    ...(miscStyles ? parseStyleProps(miscStyles, theme.mq, theme.type) : []),
  ],
});

const StyledLabel = createComponent(labelStyle, 'label', [
  'htmlFor',
  'onClick',
]);

function setVariant(
  prop,
  variant,
  getColor,
  boxModel,
  isError,
  isFocused,
  theme,
  isDisabled,
  isLabelUp
) {
  const focusedStyle = {
    backgroundColor: getColor('input', `${variant}FocusBg`),
    borderColor: isError
      ? getColor('input', `${variant}ErrorBorder`)
      : getColor('input', `${variant}FocusBorder`),
  };

  return {
    color: getColor('input', `${variant}Text`),
    backgroundColor: getColor('input', `${variant}Bg`),
    ...border(
      `${theme.inputStyle.borderWidth}px`,
      boxModel && boxModel.vp ? boxModel.vp : theme.inputStyle.lines,
      theme.inputStyle.borderStyle,
      getColor('input', `${variant}Border`)
    ),
    ...(isFocused
      ? focusedStyle
      : {
        ':hover': {
          ...(!isLabelUp
            ? { backgroundColor: getColor('input', `${variant}HoverBg`), }
            : {}),
          borderColor: getColor('input', `${variant}HoverBorder`),
          color: getColor('input', `${variant}HoverText`),
          ...(isDisabled ? { cursor: 'not-allowed', } : {}),
        },
      }),
  };
}

const labelTextStyle = ({
  labelHidden,
  isContentEditable,
  isDisabled,
  isError,
  isFocused,
  isInputEmpty,
  isTextArea,
  theme,
  variant,
}) => ({
  ...(labelHidden
    ? { display: 'none', }
    : {
      whiteSpace: 'nowrap',
      marginInlineEnd: '1rem',
      color: isError
        ? theme.color('input', `${variant}ErrorTextLabel`)
        : isDisabled
          ? theme.color('input', `${variant}TextLabelDisabled`)
          : theme.color('input', `${variant}TextLabel`),
      fontWeight: theme.inputStyle.fontWeightLabel,
      transitionProperty: 'transform',

      extend: [
        {
          condition: isTextArea === true,
          style: {
            display: 'block',
          },
        },
        {
          condition: !isTextArea && !isContentEditable,
          style: {
            bottom: '0.7rem',
            paddingRight: '1rem',
            paddingLeft: '1rem',
            marginInlineStart: '-1rem',
            ...(isFocused || !isInputEmpty
              ? {
                position: 'absolute',
                backgroundColor: theme.color(
                  'input',
                  isFocused ? `${variant}FocusBg` : `${variant}Bg`
                ),
                transform: 'translateY(-3.3rem) scale(.8)',
                paddingTop: '0.5rem',
              }
              : {}),
          },
        },
        theme.getTransition(-1, 'swiftOut'),
        {
          condition: isTextArea === true || isContentEditable === true,
          style: {
            flexGrow: 1,
            marginBottom: '1rem',
            ...borderBottom(
              `${theme.inputStyle.borderWidth}px`,
              theme.inputStyle.lines,
              theme.inputStyle.borderStyle,
              isFocused
                ? theme.color('input', `${variant}BorderTextLabel`)
                : 'transparent'
            ),
          },
        },
      ],
    }),
});

const StyledLabelText = createComponent(labelTextStyle, 'span', [ 'id', ]);

const labelAndButtonsWrapperStyle = ({ isContentEditable, isTextArea, }) => ({
  ...(isTextArea || isContentEditable ? { width: '100%', } : {}),
  ...(isContentEditable
    ? { display: 'flex', justifyContent: 'space-between', }
    : {}),
});

const StyledLabelAndButtonsWrapper = createComponent(
  labelAndButtonsWrapperStyle
);

class TextInput extends Component {
  static propTypes = {
    /**
     * An object of attrbutes to set on the DOM element.
     * Passed to the underlying input/textarea element in this component
     */
    attrs: attrsPropType,
    /** Class(es) to be added to the DOM element.
     * Automatically generated by Fela, do not enter manually.
     */
    /**
     * An object of vertical and horizontal padding inside the `<TextInput />`,
     */
    boxModel: PropTypes.shape({
      /** horizontal padding */
      hp: PropTypes.number,
      /** vertical padding */
      vp: PropTypes.number,
    }),
    className: PropTypes.string,
    /**
     * The initial value of an uncontrolled `<TextInput />`.
     * Only relevant when using an uncontrolled input
     * Allows specifying the initial value but leaving subsequent
     * updates uncontrolled
     */
    defaultValue: PropTypes.string,
    /** error note to display if input is passed a `isError` prop, */
    errorText: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.element,
      PropTypes.node,
    ]),
    /**
     * A callback function to allow parent component to get ref of input,
     * example use case: focusing the input.
     */
    refFunc: PropTypes.func,
    /**
     * Id used to connect the label to input with htmlFor for a11y reasons,
     * default will generate random id
     */
    inputId: PropTypes.string,
    /**
     * Enables rich text capabilities by using a div with
     * a `contenteditable` attribute. Should not be used
     * in cunjunction with `isTextArea`
     */
    isContentEditable: PropTypes.bool,
    /** Is this input disabled */
    isDisabled: PropTypes.bool,
    /** Is this input in error state */
    isError: PropTypes.bool,
    /**
     * Is this a textarea (multi line text input).
     * Should not be used in conjunction with `isContentEditable`
     */
    isTextArea: PropTypes.bool,
    /** label text */
    label: PropTypes.string.isRequired,
    /**
     * Used when a label is needed only for a11y, and should be visually hidden
     * in this case you should write a descriptive label that will help users
     * with screen readers understand the purpose of the input.
     * This will hide the label and add an aria-label attribute on the dom element.
     */
    labelHidden: PropTypes.bool,
    /**
     * ID used to connect the label with contentEditable div with
     * aria-describedby for a11y reasons, default will generate random id
     */
    labelId: PropTypes.string,
    /**
     * The max number of characters allowed in the input.
     *
     * Adds a native html maxlength atrribute.
     *
     * Does not work for contenteditable, if needed, should be implemented
     * via the controlling component
     */
    maxLength: PropTypes.number,
    /**
     * The min number of characters allowed in the input
     *
     * Adds a native html minlength atrribute.
     *
     * Does not work for contenteditable, if needed, should be implemented
     * via the controlling component
     */
    minLength: PropTypes.number,
    /**
     * Id used to connect the note to input with aria-describedby for a11y reasons,
     * default will generate random id
     */
    noteId: PropTypes.string,
    /** Note explaining the TextInput field  */
    noteText: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.element,
      PropTypes.node,
    ]),
    /**
     * A callback that gets called when the input blurs
     * @param {SyntheticEvent} evt - The event object
     */
    onBlur: PropTypes.func,
    /**
     * A callback that gets the new value of Input
     * used to update state of parent when using as react controlled input
     * @param {SyntheticEvent} evt - The event object
     */
    onChange: PropTypes.func,
    /**
     * A callback that gets the event triggered by onInput and the new innerHTML
     * value of the contenteditable div.
     *
     * Used to update state of parent when using as react controlled input.
     *
     * Should not be used in conjucntion with `onInput` or `onChange`
     * @param {SyntheticEvent} evt - The event object
     * @param {String} innerHtml - The new innerHtml value of the div
     */
    onContentEditableChange: PropTypes.func,
    /**
     * A callback that gets called when the input focuses.
     *
     * Used to update state of parent if the input is focused.
     * @param {SyntheticEvent} evt - The event object
     */
    onFocus: PropTypes.func,
    /**
     * Function that gets the new value of Input, alternative to onChange when needed.
     *
     * Used to update state of parent when using as react controlled input
     * @param {SyntheticEvent} evt - The event object
     */
    onInput: PropTypes.func,
    /** placeholder text */
    placeholder: PropTypes.string,
    /**
     * Used to mark an input as required, adds an abbr html element after the label
     */
    requiredText: PropTypes.shape({
      /** is the abbr wrapped in `<sup>` tags */
      isSup: PropTypes.bool,
      /** long string will show when hovering over the Abbr component */
      long: PropTypes.string.isRequired,
      /** the short string will show after the label text */
      short: PropTypes.string.isRequired,
      /**
       * miscStyles of the Abbr component
       * A special property holding miscellaneous CSS values that
       * trump all default values. Processed by
       * [`parseStyleProps`](https://Haaretz.github.io/htz-frontend/htz-css-tools#parsestyleprops)
       */
      miscStyles: stylesPropType,
    }),
    /**
     * The html input type
     * oneOf('email', 'number', 'password', 'search', 'tel', 'text', 'url')
     */
    type: PropTypes.oneOf([
      'email',
      'number',
      'password',
      'search',
      'tel',
      'text',
      'url',
    ]),
    /**
     * Value of a controlled `<TextInput />`.
     * Should never be passed manually by the consumer, but rather
     * set by the controlling component.
     */
    value: PropTypes.string,

    // Stylistic props
    /**
     * The height of an input, in rems. Only relevant when
     * `isTextArea` or `isContentEditable`.
     */
    height: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.arrayOf(
        PropTypes.shape({
          ...responsivePropBaseType,
          value: PropTypes.number.isRequired,
        })
      ),
    ]),
    /** The `<TextInput />`'s stylistic variant */
    variant: PropTypes.oneOfType([
      textInputVariantType,
      PropTypes.arrayOf(
        PropTypes.shape({
          ...responsivePropBaseType,
          value: textInputVariantType.isRequired,
        })
      ),
    ]),
    /**
     * A special property holding miscellaneous CSS values that
     * trump all default values. Processed by
     * [`parseStyleProps`](https://Haaretz.github.io/htz-frontend/htz-css-tools#parsestyleprops)
     */
    miscStyles: stylesPropType,
  };
  static defaultProps = {
    attrs: null,
    boxModel: null,
    className: null,
    defaultValue: null,
    errorText: null,
    labelHidden: false,
    height: null,
    refFunc: undefined,
    inputId: null,
    isContentEditable: false,
    isDisabled: false,
    isError: false,
    isTextArea: false,
    labelId: null,
    maxLength: null,
    minLength: null,
    miscStyles: null,
    noteId: null,
    noteText: null,
    onBlur: null,
    onChange: null,
    onContentEditableChange: null,
    onInput: null,
    onFocus: null,
    placeholder: null,
    requiredText: null,
    type: 'text',
    value: undefined,
    variant: 'primary',
  };
  static getDerivedStateFromProps(props, state) {
    return {
      isInputEmpty: !props.defaultValue && !props.value,
    };
  }
  state = {
    boldActive: false,
    disableBlur: false,
    italicActive: false,
    inputId: this.props.inputId || Math.random().toString(),
    isFocused: false,
    isInputEmpty: !this.props.defaultValue && !this.props.value,
    labelId: this.props.labelId || Math.random().toString(),
    noteId: this.props.noteId
      ? this.props.noteId
      : this.props.errorText || this.props.noteText
        ? Math.random().toString()
        : null,
  };

  /**
   * handles updating the state with the current active format buttons,
   * gets called on each keyup and mouseup while editing a contenteditable div so the format buttons
   * active state will stay acurrate.
   */
  setFormatButtonsState = () => {
    const boldActive = document.queryCommandState('bold');
    const italicActive = document.queryCommandState('italic');
    this.setState({ boldActive, italicActive, });
  };

  /**
   * handels executing clicking bold or italic format buttons
   * execCommand method allows one to run commands to manipulate the contents of the editable region
   *
   * @param {aCommandName} command - in our use case either `"bold"` or `"italic"`
   */
  toggleCommand(command) {
    document.execCommand(command);
    this.handleInputFocus(true, true);
    this.setFormatButtonsState();
  }
  /**
   * handles updating the state isFocused and disableBlur properties
   *
   * disabling the blur is needed when clicking the format buttons,
   * otherwise the contenteditable div will lose focus, change style and lose the caret position
   *
   * @param {Boolean} isFocused - Is the input focused.
   * @param {Boolean} disableFocusChange - should the focus change be disabled
   */
  handleInputFocus = (isFocused, disableFocusChange = false) => {
    if (disableFocusChange) {
      this.setState({ disableBlur: true, });
    }
    if (!this.state.disableBlur && !disableFocusChange) {
      this.setState({ isFocused, });
    }
    if (!disableFocusChange) this.setState({ disableBlur: false, });
  };
  /**
   * Used to simulate input native behaviour with contenteditable div
   * when disabled, click wont have an effect.
   *
   * Otherwise, focus on the contenteditable div (used by label and format buttons)
   *
   * @param {SyntheticEvent} evt - click event
   */
  handleClick(evt) {
    if (this.props.isDisabled) {
      evt.preventDefault();
      return;
    }
    this.contentEditableEl.focus();
  }

  render() {
    const {
      attrs,
      boxModel,
      defaultValue,
      errorText,
      labelHidden,
      height,
      noteText,
      isContentEditable,
      isError,
      isTextArea,
      isDisabled,
      label,
      maxLength,
      minLength,
      miscStyles,
      onBlur,
      onChange,
      onContentEditableChange,
      onFocus,
      onInput,
      placeholder,
      refFunc,
      requiredText,
      type,
      value,
      variant,
    } = this.props;

    const isLabelUp =
      !isTextArea &&
      !isContentEditable &&
      (this.state.isFocused || !this.state.isInputEmpty);

    return (
      <div>
        <StyledLabel
          htmlFor={this.state.inputId}
          variant={variant}
          miscStyles={miscStyles}
          boxModel={boxModel}
          isError={isError}
          isFocused={this.state.isFocused}
          isContentEditable={isContentEditable}
          isTextArea={isTextArea}
          isDisabled={isDisabled}
          isLabelUp={isLabelUp}
          {...(isContentEditable
            ? { onClick: evt => this.handleClick(evt), }
            : {})}
        >
          <StyledLabelAndButtonsWrapper
            isContentEditable={isContentEditable}
            isTextArea={isTextArea}
          >
            <StyledLabelText
              labelHidden={labelHidden}
              isContentEditable={isContentEditable}
              isDisabled={isDisabled}
              isError={isError}
              isFocused={this.state.isFocused}
              isInputEmpty={this.state.isInputEmpty}
              isTextArea={isTextArea}
              variant={variant}
              {...(isContentEditable ? { id: this.state.labelId, } : {})}
            >
              {label}
              {requiredText ? (
                <Abbr
                  miscStyles={requiredText.miscStyles}
                  requiredText={{
                    long: requiredText.long,
                    short: requiredText.short,
                  }}
                  variant={variant}
                />
              ) : null}
            </StyledLabelText>
            {isContentEditable ? (
              <div>
                {' '}
                <Button
                  variant={
                    this.state.italicActive
                      ? 'primaryOpaque'
                      : 'formattingOpaque'
                  }
                  attrs={{
                    'aria-checked': this.state.italicActive,
                    'arial-label': 'italic',
                    onMouseDown: () => this.toggleCommand('italic'),
                    role: 'switch',
                  }}
                  onClick={evt => this.handleClick(evt)}
                  boxModel={{ hp: 0, vp: 0, }}
                  miscStyles={{
                    display: this.state.isFocused ? 'initial' : 'none',
                    fontStyle: 'italic',
                    height: '5rem',
                    marginInlineEnd: '1rem',
                    width: '5rem',
                  }}
                >
                  <IconItalic />
                </Button>
                <Button
                  variant={
                    this.state.boldActive ? 'primaryOpaque' : 'formattingOpaque'
                  }
                  attrs={{
                    'aria-checked': this.state.boldActive,
                    'arial-label': 'bold',
                    onMouseDown: () => this.toggleCommand('bold'),
                    role: 'switch',
                  }}
                  onClick={evt => this.handleClick(evt)}
                  boxModel={{ hp: 0, vp: 0, }}
                  miscStyles={{
                    display: this.state.isFocused ? 'initial' : 'none',
                    height: '5rem',
                    // marginInlineEnd: '1rem',
                    marginTop: 'calc(1px - 1rem)',
                    width: '5rem',
                  }}
                >
                  <IconBold />
                </Button>
              </div>
            ) : null}
          </StyledLabelAndButtonsWrapper>
          <InputElement
            ariaDescribedBy={this.state.noteId}
            ariaLabel={labelHidden ? label : null}
            attrs={attrs}
            defaultValue={defaultValue}
            height={isTextArea || isContentEditable ? height : null}
            inputId={this.state.inputId}
            isContentEditable={isContentEditable}
            isDisabled={isDisabled}
            isFocused={this.state.isFocused}
            isTextArea={isTextArea}
            {...(isContentEditable ? { labelId: this.state.labelId, } : {})}
            {...(maxLength ? { maxLength, } : {})}
            {...(minLength ? { minLength, } : {})}
            onBlur={() => {
              this.handleInputFocus(false);
              if (onBlur) {
                onBlur();
              }
            }}
            {...(!isContentEditable
              ? {
                  onChange: evt => {
                    if (onChange) onChange(evt);
                    if (!isTextArea) {
                      if (evt.target.value.length > 0) {
                        this.setState({ isInputEmpty: false, });
                      }
                      if (evt.target.value.length === 0) {
                        this.setState({ isInputEmpty: true, });
                      }
                    }
                  },
                }
              : {})}
            onFocus={() => {
              this.handleInputFocus(true);
              if (onFocus) {
                onFocus();
              }
            }}
            {...(onInput ? { onInput, } : {})}
            placeholder={placeholder}
            refFunc={el => {
              this.contentEditableEl = el;
              if (refFunc) refFunc(el);
            }}
            {...(isContentEditable
              ? {
                  setFormatButtonsState: this.setFormatButtonsState,
                  onContentEditableChange,
                }
              : {})}
            type={type}
            value={value}
          />
        </StyledLabel>
        <Note
          text={isError ? errorText : noteText}
          isError={isError}
          noteId={this.state.noteId}
          variant={variant}
        />
      </div>
    );
  }
}

export default TextInput;
