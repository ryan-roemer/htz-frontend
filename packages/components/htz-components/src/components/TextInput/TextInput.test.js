import React from 'react';
import felaSnapshotter from '../../test-helpers/felaSnapshotter';
import TextInput from './TextInput';

describe('InputTest', () => {
  it('should correctly render a input component with default style', () => {
    const snapshot = felaSnapshotter(<TextInput />);
    expect(snapshot).toMatchSnapshot();
  });
  it('should correctly render a input component with error state style', () => {
    const snapshot = felaSnapshotter(<TextInput isError />);
    expect(snapshot).toMatchSnapshot();
  });
  it('render the component as a "textarea" element when the "isTextArea" prop is passed', () => {
    const snapshot = felaSnapshotter(<TextInput isTextArea />);
    expect(snapshot).toMatchSnapshot();
  });
  it('render the component as a input element with a custom yellow border color when isError is false', () => {
    const snapshot = felaSnapshotter(
      <TextInput inputBorderColor="yellow" inputBorderColorError="purple" />
    );
    expect(snapshot).toMatchSnapshot();
  });
});
it('render the component as a input element with a custom purple border color when isError', () => {
  const snapshot = felaSnapshotter(
    <TextInput isError inputBorderColor="yellow" inputBorderColorError="purple" />
  );
  expect(snapshot).toMatchSnapshot();
});
it('correctly render element and style based on styleObj prop', () => {
  const snapshot = felaSnapshotter(
    <TextInput
      styleObject={{
        height: '7em',
        width: '70%',
        borderWidth: '5px',
        color: 'purple',
        borderStyle: 'dashed',
      }}
    />
  );

  expect(snapshot).toMatchSnapshot();
});
