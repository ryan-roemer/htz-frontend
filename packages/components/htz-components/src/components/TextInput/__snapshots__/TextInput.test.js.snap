// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputTest render the component as a "textarea" element when the "isTextArea" prop is passed 1`] = `
Object {
  "component": <div>
    <textarea
      className="htz-a htz-b htz-c"
    />
  </div>,
  "styles": ".htz-a{border-color:green}.htz-b{height:10rem}.htz-c{resize:none}",
}
`;

exports[`InputTest render the component as a input element with a custom yellow border color when isError is false 1`] = `
Object {
  "component": <div>
    <input
      className="htz-a htz-b htz-c"
    />
  </div>,
  "styles": ".htz-a{border-color:yellow}.htz-b{height:3rem}.htz-c{resize:none}",
}
`;

exports[`InputTest should correctly render a input component with default style 1`] = `
Object {
  "component": <div>
    <input
      className="htz-a htz-b htz-c"
    />
  </div>,
  "styles": ".htz-a{border-color:green}.htz-b{height:3rem}.htz-c{resize:none}",
}
`;

exports[`InputTest should correctly render a input component with error state style 1`] = `
Object {
  "component": <div>
    <input
      className="htz-a htz-b htz-c"
    />
  </div>,
  "styles": ".htz-a{border-color:red}.htz-b{height:3rem}.htz-c{resize:none}",
}
`;

exports[`correctly render element and style based on styleObj prop 1`] = `
Object {
  "component": <div>
    <input
      className="htz-a htz-b htz-c htz-d htz-e htz-f htz-g"
    />
  </div>,
  "styles": ".htz-a{border-color:green}.htz-b{height:7em}.htz-c{resize:none}.htz-d{width:70%}.htz-e{border-width:5px}.htz-f{color:purple}.htz-g{border-style:dashed}",
}
`;

exports[`render the component as a input element with a custom purple border color when isError 1`] = `
Object {
  "component": <div>
    <input
      className="htz-a htz-b htz-c"
    />
  </div>,
  "styles": ".htz-a{border-color:purple}.htz-b{height:3rem}.htz-c{resize:none}",
}
`;
